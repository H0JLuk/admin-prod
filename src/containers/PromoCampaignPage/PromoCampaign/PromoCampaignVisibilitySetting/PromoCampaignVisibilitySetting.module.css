.page {
    position: relative;
    display: flex;
    flex-direction: column;
    height: 100%;
}

.tableWrapper {
    padding: 20px;
    background-color: #f0f2f5;
    flex-grow: 1;
}

.checked {
    font-weight: bold;
}

.checked > button {
    margin-left: 10px;
}

.footer {
    position: sticky;
    bottom: 0;
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: auto;
    padding: 20px;
    background-color: #ffffff;
    min-height: 72px;
}
