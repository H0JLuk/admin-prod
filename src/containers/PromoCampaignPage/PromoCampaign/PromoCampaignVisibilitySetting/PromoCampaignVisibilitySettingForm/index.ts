import PromoCampaignVisibilitySettingForm from './PromoCampaignVisibilitySettingForm';

export default PromoCampaignVisibilitySettingForm;
