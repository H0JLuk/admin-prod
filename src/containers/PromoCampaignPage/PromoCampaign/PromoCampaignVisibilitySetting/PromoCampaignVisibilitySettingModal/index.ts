import PromoCampaignVisibilitySettingModal from './PromoCampaignVisibilitySettingModal';

export default PromoCampaignVisibilitySettingModal;
