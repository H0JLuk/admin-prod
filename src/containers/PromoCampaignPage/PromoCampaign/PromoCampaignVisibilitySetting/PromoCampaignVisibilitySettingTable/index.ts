import PromoCampaignVisibilitySettingTable from './PromoCampaignVisibilitySettingTable';

export default PromoCampaignVisibilitySettingTable;
