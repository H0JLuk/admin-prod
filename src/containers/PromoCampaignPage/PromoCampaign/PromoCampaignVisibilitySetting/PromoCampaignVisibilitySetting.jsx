import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { useParams, useRouteMatch, useHistory } from 'react-router-dom';
import PromoCampaignVisibilitySettingTable
    from './PromoCampaignVisibilitySettingTable/PromoCampaignVisibilitySettingTable';
import Button from '../../../../components/Button/Button';

import styles from './PromoCampaignVisibilitySetting.module.css';
import {
    deletePromoCampaignVisibilitySetting, editPromoCampaignVisibilitySetting,
    getPromoCampaignVisibilitySettings
} from '../../../../api/services/promoCampaignService';

const defaultParams = {
    pageNo: 0,
    pageSize: 10,
    sortBy: 'id',
    direction: 'ASC',
    totalElements: 0,
};

// eslint-disable-next-line
const getURLSearchParams = ({ totalElements, ...rest }) => new URLSearchParams(rest).toString();

function PromoCampaignVisibilitySetting() {
    const match = useRouteMatch();
    const history = useHistory();
    const { promoCampaignId } = useParams();
    const [ selectedSettings, setSelectedSettings ] = useState(null);
    const [ loading, setLoading ] = useState(false);
    const [ visibilitySettings, setVisibilitySettings ] = useState([]);
    const [ params, setParams] = useState(defaultParams);

    const loadData = useCallback(async (searchParams = defaultParams) => {
        const urlSearchParams = getURLSearchParams(searchParams);

        try {
            const { visibilitySettings, totalElements, pageNo } = await getPromoCampaignVisibilitySettings(promoCampaignId, urlSearchParams);

            setParams({
                ...searchParams,
                pageNo,
                totalElements,
            });

            setVisibilitySettings(visibilitySettings);
        } catch (e) {
            console.error(e);
        }
    }, [promoCampaignId]);

    useEffect(() => {
        const load = async () => {
            setLoading(true);
            await loadData();
            setLoading(false);
        };

        load();
    }, [loadData]);

    const changeVisible = useCallback(async (visibilitySetting) => {
        const { id, locationId, salePointId, visible } = visibilitySetting;

        setLoading(true);

        try {
            await editPromoCampaignVisibilitySetting(id, { promoCampaignId: Number(promoCampaignId), locationId, salePointId, visible });
            await loadData(params);
        } catch (e) {
            console.error(e);
        } finally {
            setLoading(false);
        }
    }, [params, loadData, promoCampaignId]);

    const onDelete = useCallback(async () => {
        setLoading(true);

        try {
            await Promise.all(selectedSettings.map(deletePromoCampaignVisibilitySetting));
            await loadData(params);
            setSelectedSettings(null);
        } catch (e) {
            console.error(e);
        } finally {
            setLoading(false);
        }

    }, [selectedSettings, params, loadData]);

    const onCreate = useCallback(() => {
        history.push(`${match.url}/create`);
    }, [history, match.url]);

    const onEnableSelection = useCallback(() => setSelectedSettings([]), []);
    const onDisableSelection = useCallback(() => setSelectedSettings(null), []);

    const changeParams = useCallback(async (pagination, _, sorter) => {
        setLoading(true);

        await loadData({
            ...params,
            pageNo: pagination.current - 1,
            sortBy: sorter.field || 'id',
            direction: sorter.order === 'descend' ? 'DESC' : 'ASC',
        });

        setLoading(false);
    }, [params, loadData]);

    const pagination = useMemo(() => ({
        current: params.pageNo + 1,
        total: params.totalElements,
        pageSize: params.pageSize,
        showSizeChanger: false
    }), [params.pageNo, params.pageSize, params.totalElements]);

    const rowSelection = useMemo(() => selectedSettings !== null ? {
            selectedRowKeys: selectedSettings,
            onChange: (sel) => setSelectedSettings(sel),
        } : undefined,
        [selectedSettings]);

    return (
        <div className={styles.page}>
            <div className={ styles.header } >
                <h1>Настройка видимости промо-кампании</h1>
                <div>
                    {
                        selectedSettings === null ? (
                                <>
                                    <Button type="green" label="ДОБАВИТЬ" className={ styles.button_margin } onClick={ onCreate } />
                                    <Button type="blue"  label="ВЫБРАТЬ" className={ styles.button_margin } onClick={ onEnableSelection } disabled={ loading } />
                                </>
                            ) : (
                                <Button type="blue"  label="ОТМЕНИТЬ" className={ styles.button_margin } onClick={ onDisableSelection } disabled={ loading } />
                            )
                    }
                </div>
            </div>
            <PromoCampaignVisibilitySettingTable dataSource={ visibilitySettings } onChange={ changeParams } onChangeVisible={ changeVisible } loading={ loading } rowSelection={ rowSelection } pagination={ pagination } />
            {
                selectedSettings &&
                <div className={ styles.footer }>
                    <div>ВЫБРАНО {selectedSettings.length}</div>
                    {
                        selectedSettings.length > 0 && <Button type="red" label="УДАЛИТЬ НАСТРОЙКИ" className={ styles.button_margin } onClick={ onDelete } disabled={ loading } />
                    }
                </div>
            }
        </div>
    );
}

export default PromoCampaignVisibilitySetting;