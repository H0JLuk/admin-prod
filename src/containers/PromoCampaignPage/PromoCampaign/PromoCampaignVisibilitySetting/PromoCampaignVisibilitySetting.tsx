import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { useParams, useHistory, useLocation, generatePath, useRouteMatch } from 'react-router-dom';
import HeaderWithActions from '@components/HeaderWithActions';
import PromoCampaignVisibilitySettingTable
    from './PromoCampaignVisibilitySettingTable';
import { Button, message, TablePaginationConfig, TableProps } from 'antd';
import Header from '@components/Header';
import PromoCampaignVisibilitySettingModal from './PromoCampaignVisibilitySettingModal';
import {
    deletePromoCampaignVisibilitySetting,
    editPromoCampaignVisibilitySetting,
    getPromoCampaignVisibilitySettings,
    changeVisibleOfVisibilitySettings
} from '@apiServices/promoCampaignService';
import { getPathForCreatePromoCampaignVisibititySetting } from '@utils/appNavigation';
import { getSearchParamsFromUrl } from '@utils/helper';
import { VisibilitySettingDto } from '@types';
import { ButtonProps } from '@components/HeaderWithActions';
import { DIRECTION } from '@constants/common';
import { SearchParams } from '@components/HeaderWithActions/types';

import styles from './PromoCampaignVisibilitySetting.module.css';

const defaultParams: SearchParams = {
    pageNo: 0,
    pageSize: 10,
    sortBy: 'id',
    direction: DIRECTION.ASC,
    filterText: '',
    totalElements: 0,
};

const locale = {
    items_per_page: '',
    prev_page:      'Назад',
    next_page:      'Вперед',
    jump_to:        'Перейти',
    prev_5:         'Предыдущие 5',
    next_5:         'Следующие 5',
    prev_3:         'Предыдущие 3',
    next_3:         'Следующие 3',
};

const SEARCH_SETTING = 'Поиск настройки';
const BUTTON_ADD = 'Добавить';
const HEADER_TITLE = 'Настройки видимости промо-кампании';
const BUTTON_CHOOSE = 'Выбрать';
const BUTTON_CHOOSE_ALL = 'Выбрать все';
const BUTTON_UNSELECT_ALL = 'Отменить выбор';
const BUTTON_CANCEL = 'Отмена';
const BUTTON_DELETE = 'Удалить';
const TURN_ON_ALL = 'Включить все';
const TURN_OFF_ALL = 'Выключить все';

const DROPDOWN_SORT_MENU = [
    { name: 'locationName', label: 'По локации' },
    { name: 'salePointName', label: 'По точке продажи' },
    { name: 'visible', label: 'По видимости' },
];

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const getURLSearchParams = ({ totalElements, ...rest }: Record<string, string | number>) =>
    new URLSearchParams(rest as Record<string, string>).toString();

type PromoCampaignVisibilitySettingProps = {
    searchAndSortMode: boolean;
    hideHeader: boolean;
    addNewByModal: boolean;
};

type accessSelectedSettings = VisibilitySettingDto & {
    index: number;
};

const PromoCampaignVisibilitySetting: React.FC<PromoCampaignVisibilitySettingProps> = ({ searchAndSortMode = true, hideHeader, addNewByModal }) => {
    const match = useRouteMatch();
    const { search, state } = useLocation();
    const history = useHistory();
    const { promoCampaignId } = useParams<{ promoCampaignId: string; }>();
    const [ selectedSettings, setSelectedSettings ] = useState<number[] | null>(null);
    const [ loading, setLoading ] = useState(true);
    const [ visibilitySettings, setVisibilitySettings ] = useState<VisibilitySettingDto[]>([]);
    const [ params, setParams ] = useState(defaultParams);
    const [ isModalVisible, setIsModalVisible ] = useState(false);

    const showModal = useCallback(() => setIsModalVisible(true), []);

    const closeModal = useCallback(() => setIsModalVisible(false), []);

    const loadData = useCallback(async (searchParams: SearchParams = defaultParams) => {
        setLoading(true);

        try {
            const urlSearchParams = getURLSearchParams(searchParams);
            const { visibilitySettings, pageNo, totalElements } = await getPromoCampaignVisibilitySettings(+promoCampaignId, urlSearchParams);

            history.replace(`${match.url}?${urlSearchParams}`, state);
            setParams({ ...searchParams, pageNo, totalElements });
            setVisibilitySettings(visibilitySettings);
        } catch (e) {
            console.error(e); // TODO: add error handler
        }

        setLoading(false);
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [promoCampaignId]);

    const forceUpdate = useCallback(() => loadData(params), [loadData, params]);

    useEffect(() => {
        loadData(getSearchParamsFromUrl(search, defaultParams));
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [loadData]);

    const changeVisible = useCallback(async (visibilitySetting: VisibilitySettingDto) => {
        const { id, locationId, salePointId, visible } = visibilitySetting;
        setLoading(true);

        try {
            await editPromoCampaignVisibilitySetting(id, { promoCampaignId: Number(promoCampaignId), locationId, salePointId, visible });
            setVisibilitySettings((prev) => {
                const index = prev.findIndex((setting) => setting.id === id);
                return [...prev.slice(0, index), { ...prev[index], visible }, ...prev.slice(index + 1)];
            });
        } catch (e) {
            console.error(e);
            message.error(e.message);
        }

        setLoading(false);
    }, [promoCampaignId]);

    const changeVisibleAll = useCallback(async (visible: boolean) => {
        const accessSettings = selectedSettings!.reduce<accessSelectedSettings[]>((result, settingId) => {
            const index = visibilitySettings.findIndex(({ id }) => id === settingId);
            if (visibilitySettings[index].visible !== visible) {
                return [...result, {
                    ...visibilitySettings[index],
                    visible,
                    index,
                }];
            }

            return result;
        }, []);

        setLoading(true);

        try {
            const settings = accessSettings!.reduce((setting, { id }) => ({ ...setting, [id]: visible }), {});
            await changeVisibleOfVisibilitySettings(settings);
            setVisibilitySettings((prev) => {
                const nextState = prev.slice();
                accessSettings.forEach(setting => nextState[setting.index] = { ...nextState[setting.index], visible });
                return nextState;
            });
        } catch (e) {
            message.error(e.message);
        }

        setLoading(false);
    }, [selectedSettings, visibilitySettings]);

    const onDelete = useCallback(async () => {
        setLoading(true);

        try {
            await Promise.all(selectedSettings!.map(deletePromoCampaignVisibilitySetting));
            await loadData(params);
            setSelectedSettings([]);
        } catch (e) {
            console.error(e);
        } finally {
            setLoading(false);
        }

    }, [selectedSettings, params, loadData]);

    const onCreate = useCallback(() => {
        const path = getPathForCreatePromoCampaignVisibititySetting();
        history.push(generatePath(path, { promoCampaignId }));
    }, [history, promoCampaignId]);

    const onEnableSelection = useCallback(() => setSelectedSettings([]), []);
    const onDisableSelection = useCallback(() => setSelectedSettings(null), []);

    const selectAll = useCallback(() => {
        setSelectedSettings(
            (state) => state!.length === visibilitySettings.length ? [] : visibilitySettings.map(({ id }) => id)
        );
    }, [visibilitySettings]);

    const rowSelection: TableProps<VisibilitySettingDto>['rowSelection'] = useMemo(() => selectedSettings !== null
        ? {
            selectedRowKeys: selectedSettings,
            onChange: setSelectedSettings as (keys: React.Key[], values: VisibilitySettingDto[]) => void,
        } : undefined,
    [selectedSettings]);

    const selectRow = useCallback((id) => {
        if (selectedSettings !== null) {
            setSelectedSettings(
                (state) => state!.includes(id) ? state!.filter(el => el !== id) : [...state!, id]
            );
        }
    }, [selectedSettings]);

    const onChangePage: TableProps<VisibilitySettingDto>['onChange'] = async ({ pageSize, current }) => {
        setSelectedSettings((prev) => !prev ? prev : []);
        setLoading(true);
        await loadData({
            ...params,
            pageNo: current! - 1,
            pageSize: pageSize as number,
        });
        setLoading(false);
    };

    const pagination: TablePaginationConfig = useMemo(() => ({
        current: (params.pageNo as number) + 1,
        total: params.totalElements as number,
        pageSize: params.pageSize as number,
        locale,
        showSizeChanger: true,
        showQuickJumper: true,
    }), [params.pageNo, params.pageSize, params.totalElements]);

    const buttons: ButtonProps[] = useMemo(() => {
        const isSelectAll = selectedSettings?.length === visibilitySettings.length;
        if (selectedSettings === null) {

            return [
                {
                    type: 'primary',
                    label: BUTTON_ADD,
                    onClick: addNewByModal ? showModal : onCreate,
                    disabled: loading,
                },
                {
                    label: BUTTON_CHOOSE,
                    onClick: onEnableSelection,
                    disabled: loading,
                },
            ];
        }

        return [
            {
                type: 'primary',
                label:  isSelectAll ? BUTTON_UNSELECT_ALL : BUTTON_CHOOSE_ALL,
                onClick: selectAll, disabled: loading,
            },
            { label: BUTTON_CANCEL, onClick: onDisableSelection, disabled: loading },
        ];
    }, [
        selectedSettings,
        loading,
        onEnableSelection,
        selectAll,
        onDisableSelection,
        visibilitySettings,
        addNewByModal,
        onCreate,
        showModal
    ]);

    const onChangeSort = useCallback(() => {
        setSelectedSettings((state) => !state ? null : []);
    }, []);

    return (
        <>
            <div className={styles.page}>
                {!hideHeader && <Header />}
                <HeaderWithActions
                    title={HEADER_TITLE}
                    buttons={buttons}
                    showSorting={searchAndSortMode}
                    showSearchInput={searchAndSortMode && selectedSettings === null}
                    params={params}
                    setParams={setParams}
                    loadData={loadData}
                    onChangeSort={onChangeSort}
                    loading={loading}
                    inputPlaceholder={SEARCH_SETTING}
                    menuItems={DROPDOWN_SORT_MENU}
                    enableAsyncSort
                />
                <div className={styles.tableWrapper}>
                    <PromoCampaignVisibilitySettingTable
                        dataSource={visibilitySettings}
                        selectRow={selectRow}
                        onChangeVisible={changeVisible}
                        loading={loading}
                        rowSelection={rowSelection}
                        pagination={pagination}
                        onChange={onChangePage}
                    />
                </div>
                {selectedSettings && (
                    <div className={styles.footer}>
                        <div className={styles.checked}>
                            {`Выбрано настроек: ${ selectedSettings.length }`}
                            {!!selectedSettings.length && (
                                <>
                                    <Button
                                        type="primary"
                                        onClick={() => changeVisibleAll(true)}
                                    >
                                        {TURN_ON_ALL}
                                    </Button>
                                    <Button onClick={() => changeVisibleAll(false)}>
                                        {TURN_OFF_ALL}
                                    </Button>
                                </>
                            )}
                        </div>
                        {!!selectedSettings.length && (
                            <Button
                                onClick={onDelete}
                                disabled={loading}
                                type="primary"
                                danger
                            >
                                {BUTTON_DELETE}
                            </Button>
                        )}
                    </div>
                )}
            </div>
            {addNewByModal && (
                <PromoCampaignVisibilitySettingModal
                    forceUpdate={forceUpdate}
                    promoCampaignId={+promoCampaignId}
                    closeModal={closeModal}
                    isModalVisible={isModalVisible}
                />
            )}
        </>
    );
};

export default PromoCampaignVisibilitySetting;
