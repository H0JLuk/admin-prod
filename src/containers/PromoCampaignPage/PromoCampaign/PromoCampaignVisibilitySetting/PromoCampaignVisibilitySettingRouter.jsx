import React, { useCallback } from 'react';
import { Redirect, Route, Switch, useRouteMatch, useHistory } from 'react-router-dom';
import PromoCampaignVisibilitySetting from './PromoCampaignVisibilitySetting';
import PromoCampaignVisibilitySettingForm
    from './PromoCampaignVisibilitySettingForm/PromoCampaignVisibilitySettingForm';

const PromoCampaignVisibilitySettingRouter = () => {
    const match = useRouteMatch();
    const history = useHistory();

    const redirectToVisibilitySetting = useCallback(() => {
        history.push(`${match.url}`);
    }, [history, match.url]);

    return (
        <Switch>
            <Route exact path={ `${match.path}` } component={ PromoCampaignVisibilitySetting } />
            <Route
                path={ `${match.path}/create` }
                render={ routeProps => (
                    <PromoCampaignVisibilitySettingForm
                        { ...routeProps }
                        onSubmit={ redirectToVisibilitySetting }
                        onCancel={ redirectToVisibilitySetting }
                    />
                ) }
            />
            <Redirect to={ `${match.path}` } />
        </Switch>
    );
};

export default PromoCampaignVisibilitySettingRouter;
