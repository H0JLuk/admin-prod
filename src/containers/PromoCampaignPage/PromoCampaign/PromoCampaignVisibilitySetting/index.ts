import PromoCampaignVisibilitySetting from './PromoCampaignVisibilitySetting';

export default PromoCampaignVisibilitySetting;
