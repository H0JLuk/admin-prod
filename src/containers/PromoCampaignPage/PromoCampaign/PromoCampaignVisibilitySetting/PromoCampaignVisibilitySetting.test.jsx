import React from 'react';
import { generatePath } from 'react-router-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { shallow } from 'enzyme';
import { message } from 'antd';

import PromoCampaignVisibilitySetting from './PromoCampaignVisibilitySetting';
import { act } from 'react-dom/test-utils';
import { changeVisibleOfVisibilitySettings, deletePromoCampaignVisibilitySetting, editPromoCampaignVisibilitySetting, getPromoCampaignVisibilitySettings } from '../../../../api/services/promoCampaignService';
import { getSearchParamsFromUrl } from '../../../../utils/helper';
import { visibilitySettingTestData, promoCampaignTestData } from '../../../../../__tests__/constants';
import { getPathForCreatePromoCampaignVisibititySetting } from '../../../../utils/appNavigation';
import { sleep } from '../../../../setupTests';


const defaultParams = {
    pageNo: 0,
    pageSize: 10,
    sortBy: 'id',
    direction: 'ASC',
    filterText: '',
    totalElements: 0,
};

jest.mock('../../../../components/Header/Header', () => () => (<div id="#header">Header</div>));

jest.mock('../../../../utils/appNavigation', () => ({
    getPathForCreatePromoCampaignVisibititySetting: jest.fn(),
}));

jest.mock('../../../../api/services/promoCampaignService', () => ({
    getPromoCampaignVisibilitySettings: jest.fn(),
    editPromoCampaignVisibilitySetting: jest.fn(),
    changeVisibleOfVisibilitySettings: jest.fn(),
    deletePromoCampaignVisibilitySetting: jest.fn(),
}));

jest.mock('../../../../utils/helper', () => ({
    getSearchParamsFromUrl: jest.fn(),
}));

jest.mock('antd', () => ({
    ...jest.requireActual('antd'),
    message: {
        error: jest.fn(),
    }
}));

const mockHistoryPush = jest.fn();
const mockHistoryReplace = jest.fn();
const mockPromoCampaign = promoCampaignTestData;
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    generatePath: jest.fn(),
    useHistory: () => ({
        push: mockHistoryPush,
        replace: mockHistoryReplace,
        location: {
            search: '?pageNo=0&pageSize=10&sortBy=id&direction=ASC&filterText=',
        },
    }),
    useRouteMatch: () => ({
        url: '/admin/promo-campaign/24/edit',
    }),
    useParams: () => ({
        promoCampaignId: '24',
    }),
    useLocation: () => ({
        search: '?pageNo=0&pageSize=10&sortBy=id&direction=ASC&filterText=',
        state: {
            promoCampaign: mockPromoCampaign,
        },
    }),
}));

const urlSearchParams = 'pageNo=0&pageSize=10&sortBy=id&direction=ASC&filterText=';

const getSearchParamsFromUrlResult = {
    pageNo: '0',
    pageSize: '10',
    sortBy: 'id',
    direction: 'ASC',
    filterText: '',
    totalElements: 0,
};

beforeEach(() => {
    getPromoCampaignVisibilitySettings.mockResolvedValue(visibilitySettingTestData);
    getSearchParamsFromUrl.mockReturnValue(getSearchParamsFromUrlResult);
});

describe('<PromoCampaignVisibilitySetting /> tests', () => {
    const props = {
        searchAndSortMode: false,
        hideHeader: false,
        addNewByModal: true,
    };

    it('PromoCampaignVisibilitySetting snapshot', () => {
        const wrapper = shallow(<PromoCampaignVisibilitySetting { ...props } />);
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should run loadData', async () => {
        await act(async () => {
            render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            expect(getSearchParamsFromUrl).toBeCalledWith(`?${urlSearchParams}`, defaultParams);
            expect(getPromoCampaignVisibilitySettings).toBeCalledWith('24', urlSearchParams);
            expect(mockHistoryReplace).toBeCalled();
        });
    });

    it('should reject loadData', async () => {
        const spy = jest.spyOn(console, 'error').mockImplementation(() => '');
        getPromoCampaignVisibilitySettings.mockRejectedValueOnce('error');
        await act(async () => {
            render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            expect(spy).toHaveBeenCalled();
        });
    });

    it('should run enableSelection / disableSelection', async () => {
        await act(async () => {
            const { getByText } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            fireEvent.click(getByText(/Выбрать/));
            expect(getByText(/Выбрано настроек: 0/)).toBeInTheDocument();
            fireEvent.click(getByText(/Отмена/));
            expect(getByText(/Выбрать/)).toBeInTheDocument();
        });
    });

    it('should run changeVisible on switch click', async () => {
        editPromoCampaignVisibilitySetting.mockResolvedValue('');
        await act(async () => {
            const { container } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            const switchBtn = container.querySelector('.ant-switch');
            fireEvent.click(switchBtn);
            expect(editPromoCampaignVisibilitySetting).toBeCalledTimes(1);
        });
    });

    it('should render error message on changeVisible run', async () => {
        editPromoCampaignVisibilitySetting.mockRejectedValueOnce('error');
        const spy = jest.spyOn(console, 'error').mockImplementation(() => '');
        await act(async () => {
            const { container } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            const switchBtn = container.querySelector('.ant-switch');
            fireEvent.click(switchBtn);
            await sleep();
            expect(message.error).toBeCalled();
            expect(spy).toHaveBeenCalled();
        });
    });

    it('should run changeVisibleAll', async () => {
        changeVisibleOfVisibilitySettings.mockResolvedValue('');
        await act(async () => {
            const { getByText } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(getByText(/Выбрать все/));
            fireEvent.click(getByText(/Включить все/));
            expect(changeVisibleOfVisibilitySettings).toBeCalled();
        });
    });

    it('should render error message on changeVisibleAll run', async () => {
        changeVisibleOfVisibilitySettings.mockRejectedValueOnce('error');
        await act(async () => {
            const { getByText } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(getByText(/Выбрать все/));
            fireEvent.click(getByText(/Включить все/));
            fireEvent.click(getByText(/Выключить все/));
            expect(changeVisibleOfVisibilitySettings).toBeCalledTimes(2);
        });
    });

    it('should run onDelete fns', async () => {
        deletePromoCampaignVisibilitySetting.mockResolvedValue('');
        await act(async () => {
            const { getByText } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(getByText(/Выбрать все/));
            fireEvent.click(getByText(/Удалить/));
            expect(deletePromoCampaignVisibilitySetting).toBeCalledTimes(11);
            expect(getPromoCampaignVisibilitySettings).toBeCalledWith('24', urlSearchParams);
        });
    });

    it('should log error message in console after run onDelete fn', async () => {
        const spy = jest.spyOn(console, 'error').mockImplementation(() => '');
        deletePromoCampaignVisibilitySetting.mockRejectedValueOnce('error');

        await act(async () => {
            const { getByText } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            fireEvent.click(getByText(/Выбрать/));
            fireEvent.click(getByText(/Выбрать все/));
            fireEvent.click(getByText(/Удалить/));
            await sleep();
            expect(spy).toBeCalled();
        });
    });

    it('should run onCreate fn', async () => {
        const path = '/admin/promo-campaign/:promoCampaignId/visibility-setting/create';
        const generatePathResult = path.replace(':promoCampaignId', '24');
        const newProps = { ...props, addNewByModal: false };

        getPathForCreatePromoCampaignVisibititySetting.mockReturnValue(path);
        generatePath.mockReturnValue(generatePathResult);
        await act(async () => {
            const { getByText } = render(<PromoCampaignVisibilitySetting { ...newProps } />);
            await sleep();
            fireEvent.click(getByText(/Добавить/));
            expect(getPathForCreatePromoCampaignVisibititySetting).toBeCalled();
            expect(generatePath).toBeCalledWith(path, { promoCampaignId: '24' });
            expect(mockHistoryPush).toBeCalledWith(generatePathResult);
        });
    });

    it('should select item by click on the row', async () => {
        await act(async () => {
            const { getByText, container } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            fireEvent.click(getByText(/Выбрать/));
            const tr = container.querySelector('.ant-table-row.ant-table-row-level-0');
            const checkbox = container.querySelector('.ant-checkbox-input');
            fireEvent.click(tr);
            expect(checkbox.checked).toBe(true);
            fireEvent.click(tr);
            expect(checkbox.checked).toBe(false);
        });
    });

    it('should run load data after new page rendered', async () => {
        await act(async () => {
            const { container } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            const page2 = container.querySelectorAll('.ant-pagination-item')[1];
            fireEvent.click(page2);
            expect(getPromoCampaignVisibilitySettings).toBeCalledWith('24', urlSearchParams);
        });
    });

    it('run onChangeSort after change sort param', async () => {
        const newProps = { ...props, searchAndSortMode: true };
        await act(async () => {
            const { getByText, queryAllByText } = render(<PromoCampaignVisibilitySetting { ...newProps } />);
            await sleep();
            fireEvent.click(getByText(/Сортировать/));
            fireEvent.click(screen.getByText(/По локации/));
            expect(queryAllByText(/По локации/)[0]).toBeInTheDocument();
        });
    });

    it('should show/close modal', async () => {
        await act(async () => {
            const { getByText } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            fireEvent.click(getByText(/Добавить/));
            await sleep();
            expect(document.querySelector('.ant-modal')).toBeInTheDocument();

            fireEvent.click(document.querySelector('.ant-modal-close-x'));
            await sleep();

            expect(document.querySelector('.ant-modal')).toBe(null);
        });
    });

    it('should run forceUpdate fn', async () => {
        await act(async () => {
            const wrapper = shallow(<PromoCampaignVisibilitySetting { ...props } />);
            wrapper.find('PromoCampaignVisibilitySettingModal').prop('forceUpdate')();
            await sleep();
            expect(getPromoCampaignVisibilitySettings).toBeCalledWith('24', urlSearchParams);
        });
    });

    it('should render header', async () => {
        await act(async () => {
            const { getByText } = render(<PromoCampaignVisibilitySetting { ...props } />);
            await sleep();
            expect(getByText('Header')).toBeInTheDocument();
        });
    });

    it('should not render header', async () => {
        const newProps = { ...props, hideHeader: true };
        await act(async () => {
            const { queryAllByText } = render(<PromoCampaignVisibilitySetting { ...newProps } />);
            await sleep();
            expect(queryAllByText('Header')).toHaveLength(0);
        });
    });
});
