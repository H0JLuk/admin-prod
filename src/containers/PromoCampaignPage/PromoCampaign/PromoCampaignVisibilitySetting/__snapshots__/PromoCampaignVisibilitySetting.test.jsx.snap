// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PromoCampaignVisibilitySetting /> tests PromoCampaignVisibilitySetting snapshot 1`] = `
"<Fragment>
  <div className=\\"page\\">
    <Component />
    <HeaderWithActions title=\\"Настройки видимости промо-кампании\\" buttons={{...}} showSorting={false} showSearchInput={false} params={{...}} setParams={[Function: bound _dispatchAction]} loadData={[Function (anonymous)]} onChangeSort={[Function (anonymous)]} loading={true} inputPlaceholder=\\"Поиск настройки\\" menuItems={{...}} enableAsyncSort={true} />
    <div className=\\"tableWrapper\\">
      <PromoCampaignVisibilitySettingTable dataSource={{...}} selectRow={[Function (anonymous)]} onChangeVisible={[Function (anonymous)]} loading={true} rowSelection={[undefined]} pagination={{...}} onChange={[Function: onChangePage]} />
    </div>
  </div>
  <PromoCampaignVisibilitySettingModal forceUpdate={[Function (anonymous)]} promoCampaignId={24} closeModal={[Function (anonymous)]} isModalVisible={false} />
</Fragment>"
`;
